File: cpp/organism/animal/species/Antelope.cpp
//
// Created by Filip on 12/04/2025.
//

#include "Antelope.h"

Antelope::Antelope(Point point, World &world)
    : Animal(point, Constants::Animal::Antelope::Symbol,
             Constants::Animal::Antelope::Strength,
             Constants::Animal::Antelope::Initiative, world) {}

void Antelope::reproduce() {
    Point newPosition = world.getFreeSpace(position);
    if (newPosition == Point(-1, -1)) {
        return;
    }

    world.queueOrganismAddition(new Antelope(newPosition, world));
}

Point Antelope::getNewPosition() {
    std::vector<Point> directions = {Point(0, 1), Point(1, 0), Point(0, -1), Point(-1, 0),
                                     Point(0, 2), Point(2, 0), Point(0, -2), Point(-2, 0)};
    Point newPosition;

    int maxAttempts = directions.size();// Maksymalna liczba prób
    int attempts = 0;

    while (!directions.empty() && attempts < maxAttempts) {
        int index = std::uniform_int_distribution<>(0, directions.size() - 1)(world.getRandomEngine());
        newPosition = position + directions[index];
        directions.erase(directions.begin() + index);// Usuń sprawdzony kierunek
        attempts++;

        if (world.isInBounds(newPosition)) {
            return newPosition;
        }
    }

    // Jeśli nie znaleziono wolnej przestrzeni, zwróć aktualną pozycję
    return position;
}

bool Antelope::collision(Organism &attacker) {
    if (std::uniform_int_distribution<>(0, 1)(world.getRandomEngine()) == 0) {
        world.getLogger().logEvent("Antelope has escaped from " +
                                   attacker.getSymbol() + ".");
        move(getNewPosition());
    } else {
        return Animal::collision(attacker);// Domyślna logika kolizji
    }
    return false;
}

Organism *Antelope::clone() const {
    return new Antelope(*this);// Tworzy kopię obiektu za pomocą konstruktora kopiującego
}

File: java/src/organism/animal/species/Antelope.java
package organism.animal.species;

import constants.Constants;
import logger.EventLogger;
import organism.Organism;
import organism.animal.Animal;
import world.World;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class Antelope extends Animal {
    public Antelope(int x, int y, World world) {
        super(x, y, Constants.Animal.Antelope.Symbol, Constants.Animal.Antelope.Strength, Constants.Animal.Antelope.Initiative, world);
    }

    @Override
    protected int[] getNewPosition() {
        // Antelope moves 2 fields in one direction
        int[][] directions = {
            {0, 2}, {2, 0}, {0, -2}, {-2, 0},
            {0, 1}, {1, 0}, {0, -1}, {-1, 0},
            {2, 2}, {2, -2}, {-2, 2}, {-2, -2}, // diagonal moves (2 fields)
            {1, 1}, {1, -1}, {-1, 1}, {-1, -1}  // diagonal moves (1 field)
        };
        List<int[]> possibleMoves = new ArrayList<>();
        for (int[] dir : directions) {
            int nx = x + dir[0];
            int ny = y + dir[1];
            if (world.isInside(nx, ny)) {
                possibleMoves.add(new int[]{nx, ny});
            }
        }
        if (possibleMoves.isEmpty()) return new int[]{x, y};
        return possibleMoves.get(new Random().nextInt(possibleMoves.size()));
    }

    @Override
    public boolean collision(Organism other) {
        // 50% chance to escape
        if (new Random().nextBoolean()) {
            int[] escapePos = getNewPosition();
            if (escapePos[0] != x || escapePos[1] != y) {
                this.move(escapePos[0], escapePos[1]);
                EventLogger.getInstance().log("Antelope escaped from " + other.getClass().getName() + " at (" + x + ", " + y + ")");
                return false;
            }
        }
        return super.collision(other);
    }

    @Override
    protected void reproduce() {
        int[] newPosition = world.findNearestFree(this.getX(), this.getY());
        if (newPosition != null) {
            Animal offspring = new Antelope(newPosition[0], newPosition[1], world);
            world.addOrganism(offspring);
            EventLogger.getInstance().log("Antelope reproduced at (" + newPosition[0] + ", " + newPosition[1] + ")");
        }
    }
}